File: bench/deser_crossval.zig
const bcs = @import("bcs");
const std = @import("std");

const TinyEnum = union(enum) {
    Unit,
    Value: u64,
};

fn emitHex(name: []const u8, bytes: []const u8) void {
    std.debug.print("{s}=ok:", .{name});
    for (bytes) |b| std.debug.print("{x:0>2}", .{b});
    std.debug.print("\n", .{});
}

fn emitErr(name: []const u8) void {
    std.debug.print("{s}=err\n", .{name});
}

fn emitStatus(comptime T: type, allocator: std.mem.Allocator, name: []const u8, raw: []const u8) void {
    const value = bcs.deserialize(T, allocator, raw) catch {
        emitErr(name);
        return;
    };
    defer bcs.freeDeserialized(T, allocator, value);

    const encoded = bcs.serialize(allocator, value) catch {
        emitErr(name);
        return;
    };
    defer allocator.free(encoded);
    emitHex(name, encoded);
}

pub fn main() void {
    const allocator = std.heap.c_allocator;
    const KeySet = bcs.Map(u8, void);

    emitStatus(bool, allocator, "bool_true", &.{0x01});
    emitStatus(bool, allocator, "bool_invalid", &.{0x02});
    emitStatus(u8, allocator, "u8_trailing", &.{ 0x01, 0x02 });

    emitStatus(?u8, allocator, "opt_some_u8", &.{ 0x01, 0x2a });
    emitStatus(?u8, allocator, "opt_invalid_tag", &.{0x05});

    emitStatus(bcs.String, allocator, "string_utf8", &.{ 0x05, 'h', 'e', 'l', 'l', 'o' });
    emitStatus(bcs.String, allocator, "string_invalid_utf8", &.{ 0x02, 0xc0, 0x80 });
    emitStatus([]const u8, allocator, "bytes_invalid_utf8", &.{ 0x02, 0xc0, 0x80 });

    emitStatus(TinyEnum, allocator, "enum_unit", &.{0x00});
    emitStatus(TinyEnum, allocator, "enum_value", &.{ 0x01, 0x34, 0x12, 0, 0, 0, 0, 0, 0 });
    emitStatus(TinyEnum, allocator, "enum_invalid_tag", &.{0x02});
    emitStatus(TinyEnum, allocator, "enum_noncanonical_tag", &.{ 0x80, 0x80, 0x80, 0x80, 0x00 });

    emitStatus([]const u16, allocator, "vec_u16_ok", &.{ 0x02, 0x01, 0x00, 0x02, 0x00 });
    emitStatus([]const u16, allocator, "vec_u16_short", &.{ 0x02, 0x01, 0x00 });

    emitStatus(struct { u32, bool }, allocator, "tuple_pair", &.{ 0x2a, 0x00, 0x00, 0x00, 0x01 });

    emitStatus(KeySet, allocator, "map_empty", &.{0x00});
    emitStatus(KeySet, allocator, "map_canonical", &.{ 0x02, 0x04, 0x05 });
    emitStatus(KeySet, allocator, "map_out_of_order", &.{ 0x02, 0x05, 0x04 });
    emitStatus(KeySet, allocator, "map_duplicate", &.{ 0x02, 0x05, 0x05 });
}

File: bench/deser_crossval.rs
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Serialize, Deserialize)]
enum TinyEnum {
    Unit,
    Value(u64),
}

fn hex(bytes: &[u8]) -> String {
    bytes
        .iter()
        .map(|b| format!("{:02x}", b))
        .collect::<Vec<_>>()
        .join("")
}

fn emit<T>(name: &str, bytes: &[u8])
where
    T: Serialize + DeserializeOwned,
{
    match bcs::from_bytes::<T>(bytes) {
        Ok(value) => {
            let encoded = bcs::to_bytes(&value).unwrap();
            println!("{name}=ok:{}", hex(&encoded));
        }
        Err(_) => println!("{name}=err"),
    }
}

fn main() {
    type KeySet = BTreeMap<u8, ()>;

    emit::<bool>("bool_true", &[0x01]);
    emit::<bool>("bool_invalid", &[0x02]);
    emit::<u8>("u8_trailing", &[0x01, 0x02]);

    emit::<Option<u8>>("opt_some_u8", &[0x01, 0x2a]);
    emit::<Option<u8>>("opt_invalid_tag", &[0x05]);

    emit::<String>("string_utf8", &[0x05, b'h', b'e', b'l', b'l', b'o']);
    emit::<String>("string_invalid_utf8", &[0x02, 0xc0, 0x80]);
    emit::<Vec<u8>>("bytes_invalid_utf8", &[0x02, 0xc0, 0x80]);

    emit::<TinyEnum>("enum_unit", &[0x00]);
    emit::<TinyEnum>("enum_value", &[0x01, 0x34, 0x12, 0, 0, 0, 0, 0, 0]);
    emit::<TinyEnum>("enum_invalid_tag", &[0x02]);
    emit::<TinyEnum>("enum_noncanonical_tag", &[0x80, 0x80, 0x80, 0x80, 0x00]);

    emit::<Vec<u16>>("vec_u16_ok", &[0x02, 0x01, 0x00, 0x02, 0x00]);
    emit::<Vec<u16>>("vec_u16_short", &[0x02, 0x01, 0x00]);

    emit::<(u32, bool)>("tuple_pair", &[0x2a, 0x00, 0x00, 0x00, 0x01]);

    emit::<KeySet>("map_empty", &[0x00]);
    emit::<KeySet>("map_canonical", &[0x02, 0x04, 0x05]);
    emit::<KeySet>("map_out_of_order", &[0x02, 0x05, 0x04]);
    emit::<KeySet>("map_duplicate", &[0x02, 0x05, 0x05]);
}
